specification/sdk-exporter.md

```diff
@@ -0,0 +1,91 @@
+# Exporter Interface
+
+Exporter defines the interface that protocol-specific exporters must implement so that they can be plugged into OpenTelemetry SDK and support sending of telemetry data.
+
+The goals of the interface are:
+
+- Minimize burden of implementation for protocol-dependant telemetry exporters. The protocol exporter is expected to be primarily a simple telemetry data encoder and transmitter.
+
+- Allow implementing helpers as composable components that use the same chainable Exporter interface. SDK authors are encouraged to implement common functionality such as queuing, batching, tagging, etc. as helpers. This functionality will be applicable regardless of what protocol exporter is used.
+
+## Interface Definition
+
+The exporter must support two functions: **Export** and **Shutdown**. In strongly typed languages typically there will be 2 separate Exporter interfaces, one that accepts Spans (SpanExporter) and one that accepts Metrics (MetricsExporter).
+
+### Export(batch)
+
+Exports a batch of telemetry data. Protocol exporters that will implement this function are typically expected to serialize and transmit the data to the destination.
+
+Export() will never be called concurrently for the same exporter instance. Export() can be called again only after the current call returns.
+
+Export() must not block indefinitely, there must be a reasonable upper limit after which the call must time out with an error result(typically CannotExport).
+
+**Parameters:**
+
+batch - a batch of telemetry data. The exact data type of the batch is language specific, typically it is a list of telemetry items, e.g. for spans in Java it will be typically `List<Span>`.
+
+**Returns:** tuple of (ThrottlePeriod, ExportResult):
+
+ThrottlePeriod is duration to wait before calling Export() again. 0 duration indicates that Export() may be called immediately. Duration greater than 0 indicates that the export destination requested the sender to throttle. In this case the sender will pause sending to this particular exporter for specified duration. Note: regardless of ThrottlePeriod value, the ExportResult may be a Success or one of other values, so SDK must correctly process ExportResult value independently from ThrottlePeriod value.
```

Comment:
Not sure I understand the usefulness of the ThrottlePeriod.

Reply:
This allows destinations to signal to sources that they are overwhelmed. The source can then reduce exporting rate, e.g. by sampling. So, for example if we have a chain of exporters:

Batch -> Sampler -> Protocol Exporter ---network--> Destination Server

Destination server can signal to Protocol Exporter that it is overwhelmed, which in turn can signal to Sampler via ThrottlePeriod to begin sampling. So Sampler will be smart, it will only sample as much as it is required by Destination Server.

Reply:
Not sure if this is not Protocol specific. Don't see us implementing this.

Reply:
For best results you want intermediary helper exporters to be aware of this. Consider the example above. If throttling concept is not part of Export() function then `Sampler` has no way of acting intelligently and performing as much sampling as it is needed to satisfy the needs of the destination.

The same logic cannot be implemented if only protocol exporters are aware of throttling issues.

Anyway, I am removing this now to move forward, we can discuss throttling later.

Reply:
You can implement a Sampler that depends on this particular Exporter and control directly there. No need to return this result in the export piepline. What you suggest is possible I have an idea how.